File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Util/IFSharpElementFactory.fs
namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Util

open JetBrains.DocumentModel
open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
open JetBrains.ReSharper.Psi.Modules

type IFSharpElementFactory =
    abstract CreateOpenStatement: ns: string -> IOpenStatement
    abstract CreateWildPat: unit -> IWildPat

    abstract CreateParenExpr: unit -> IParenExpr
    abstract CreateParenExpr: IFSharpExpression -> IParenExpr

    abstract CreateConstExpr: text: string -> IConstExpr

    abstract CreateAppExpr: addSpace: bool -> IPrefixAppExpr
    abstract CreateAppExpr: funcName: string * arg: IFSharpExpression -> IPrefixAppExpr
    abstract CreateAppExpr: funExpr: IFSharpExpression * argExpr: IFSharpExpression * addSpace: bool -> IPrefixAppExpr
    abstract CreateBinaryAppExpr: string * left: IFSharpExpression * right: IFSharpExpression -> IFSharpExpression
    abstract CreateSetExpr: left: IFSharpExpression * right: IFSharpExpression -> IFSharpExpression

    abstract CreateExpr: string -> IFSharpExpression
    abstract CreateReferenceExpr: string -> IFSharpExpression
    
    abstract CreateLetBindingExpr: bindingName: string -> ILetOrUseExpr
    abstract CreateLetModuleDecl: bindingName: string -> ILetModuleDecl

    abstract CreateIgnoreApp: IFSharpExpression * newLine: bool -> IBinaryAppExpr
    abstract CreateRecordExprBinding: fieldName: string * addSemicolon: bool -> IRecordExprBinding

    abstract CreateParenPat: unit -> IParenPat
    abstract CreateTypedPat: pattern: string * typeUsage: ITypeUsage * spaceBeforeColon: bool -> ITypedPat
    
    abstract CreateTypeUsage: typeUsage: string -> ITypeUsage
    abstract CreateTypeUsage: typeUsages: ITypeUsage list -> ITypeUsage
    
    abstract CreateReturnTypeInfo: typeSignature: ITypeUsage -> IReturnTypeInfo
    
    abstract CreateMatchExpr: IFSharpExpression -> IMatchExpr
    abstract CreateMatchClause: unit -> IMatchClause
    abstract CreateForEachExpr: IFSharpExpression -> IForEachExpr

    abstract AsReferenceExpr: typeReference: ITypeReferenceName -> IReferenceExpr

    abstract CreateEmptyAttributeList: unit -> IAttributeList
    abstract CreateAttribute: attrName: string -> IAttribute


[<AllowNullLiteral>]
type IFSharpLanguageService =
    abstract CreateParser: IDocument -> IFSharpParser
    abstract CreateElementFactory: IPsiModule -> IFSharpElementFactory

File: ReSharper.FSharp/src/FSharp.Psi/src/IFSharpElementFactory.cs
using System;
using FSharp.Compiler.Symbols;
using JetBrains.ReSharper.Plugins.FSharp.Psi.Tree;
using Microsoft.FSharp.Collections;

namespace JetBrains.ReSharper.Plugins.FSharp.Psi
{
  public enum TypeUsageContext
  {
    TopLevel,
    ParameterSignature,
    Signature,
  }

  public interface IFSharpElementFactory
  {
    IOpenStatement CreateOpenStatement(string ns);

    IWildPat CreateWildPat();
    IMemberSelfId CreateSelfId(string name);

    IParenExpr CreateParenExpr();
    IConstExpr CreateConstExpr(string text);

    IPrefixAppExpr CreateAppExpr(string funcName, IFSharpExpression arg);
    IPrefixAppExpr CreateAppExpr(IFSharpExpression funExpr, IFSharpExpression argExpr, bool addSpace);
    IFSharpExpression CreateBinaryAppExpr(string opName, IFSharpExpression left, IFSharpExpression right);
    IFSharpExpression CreateSetExpr(IFSharpExpression left, IFSharpExpression right);

    IFSharpExpression CreateExpr(string expr);
    IReferenceExpr CreateReferenceExpr(string expr);

    ILetOrUseExpr CreateLetBindingExpr(string bindingName);
    ILetBindingsDeclaration CreateLetModuleDecl(string bindingName);

    IBinaryAppExpr CreateIgnoreApp(IFSharpExpression expr, bool newLine);
    IRecordFieldBinding CreateRecordFieldBinding(string fieldName, bool addSemicolon);
    IRecordFieldDeclaration CreateRecordFieldDeclaration(bool isMutable, string fieldName, ITypeUsage typeUsage);
    IFSharpPattern CreatePattern(string text, bool topLevel);
    IParenPat CreateParenPat();
    ITypedPat CreateTypedPat(IFSharpPattern pattern, ITypeUsage typeUsage);

    ITypeUsage CreateTypeUsage(string typeUsage, TypeUsageContext context);

    IReturnTypeInfo CreateReturnTypeInfo(ITypeUsage typeSignature);

    IMatchExpr CreateMatchExpr(IFSharpExpression expr);
    IMatchClause CreateMatchClause();

    IForEachExpr CreateForEachExpr(IFSharpExpression expr);

    IReferenceExpr AsReferenceExpr(ITypeReferenceName typeReference);

    IExpressionReferenceName CreateExpressionReferenceName(string referenceName);
    ITypeReferenceName CreateTypeReferenceName(string referenceName);

    IAttributeList CreateEmptyAttributeList();
    IAttribute CreateAttribute(string attrName);

    FSharpList<IParametersPatternDeclaration> CreateMemberParamDeclarations(FSharpList<FSharpList<Tuple<string, FSharpType>>> curriedParameterNames, bool isSpaceAfterComma, bool addTypes, bool preferNoParens, FSharpDisplayContext displayContext);
    IMemberDeclaration CreateMemberBindingExpr(string bindingName, FSharpList<string> typeParameters, FSharpList<IParametersPatternDeclaration> args);
    IMemberDeclaration CreatePropertyWithAccessor(string propertyName, string accessorName, FSharpList<IParametersPatternDeclaration> args);

    ITypeParameterDeclarationList CreateTypeParameterOfTypeList(FSharpList<string> names);

    IAccessorsNamesClause CreateAccessorsNamesClause(bool withGetter, bool withSetter);
  }
}
